011-kafka-canary.md

```diff
@@ -0,0 +1,40 @@
+# Strmixi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and continously producing and consuming events on the topics. 
+
+## Current situation
+
+Currently  Strimzi does not have any canary style feature. It does have readiness checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following:
+  
+  * Improved readiness check for Kafka clusters
+  * SLI(s) for service admins
+  * Metrics:
+    * End-to-end latency
+    * Succesfull/failed events
+    * Disk usage
+    * Disk errors
+
+## Proposal
+
+The canary would be implemented as part of the Kafka Roller component. 
+A single _canary_ topic would be created on creation of a Kafka cluster. The topic would have one partition for Kafka broker. Events would then be continously produced and consumed to/from the canary topic. The internal TLS 9091 port would be used to connect to the brokers.
+
+As this happens the relevant metrics would be collected and exported. Errors such as the inability to produce to the topic can then be used by the KafkaRoller to trigger a restart, acting as healthc check for the Kafka cluster. 
+
+
+## Affected/not affected projects
+
+This functionality should not affect other Strmizi projects.
+
+## Compatibility
+
+The canary would be fully backwards compatible.
+
+
+## Rejected alternatives
+
+Another option would be to create an external canary client to provide much of the above functionality. It could be deployed separately to the Strimzi operator and connect to the brokers thourgh the external route. However the benefit of building this tool into the Strimzi operator is the abilty to use it as a health check for Kafka clusters.
```

Comment:
I think that the canary as a standalone client would be better personally, and if it were useful for the operator it would be fairly trivial to pull it in as a dependency and run the libraries within the operator.

Another advantage of it being standalone is it could be used with other Kafka clusters, not just strimzi ones.

Reply:
@samuel-hawker I don't think that's unreasonable, but I think this is an opportunity to figure out what value the operator could get from this to ensure that whatever got written was actually useful for the operator use case. The value I see is:

* Deeper checking that the broker really is working, by exercising the produce and consume paths, which could be used by the KafkaRoller (we'd need to define exactly what it would do when there was a problem, but until we know the API and the errors which propagate from it that's a little hard to do). 
* Getting this information sooner after something goes wrong, perhaps triggering a reconciliation sooner than the next scheduled one.
* Some better metrics, though I think we should try to define exactly what in this proposal.

To elaborate a little.... the proposal says that it would use the internal listener. Thus any authentication exceptions from the producer would suggest that the internal PKI is broken. 

Another interesting case would be `KafkaStorageException` which is a `RetriableException` so I don't know off the top of my head how that would propagate out of a producer, but we could use that as a signal to check with `Admin.describeLogDirs`, restart the broker and prevent restarting any other brokers in the meantime. 

In both these examples the `Kafka` status could be set with a `Condition` which indicated that there was a problem and whether it was, or was not, being handled automatically.

Reply:
I think this sounds great, so the canary would be more of an API right? A standalone library that we would hook the operator straight into. Your suggestion of using it to trigger reconciliations sound useful, but I suppose that would mean spinning up a canary in a separate verticle for each operand (or indeed one canary to manage them all)?